Mathlib/AlgebraicGeometry/OpenImmersion/Basic.lean

```diff
@@ -0,0 +1,1204 @@
+    dsimp only [comp_c_app, unop_op, whiskerRight_app, pullbackConeOfLeftFst]
+    -- simp only [ofRestrict_c_app, NatTrans.comp_app]
+    simp only [Quiver.Hom.unop_op, TopCat.Presheaf.pushforwardObj_map, app_invApp_assoc,
+      eqToHom_app, eqToHom_unop, Category.assoc, NatTrans.naturality_assoc, Functor.op_map]
+    erw [← Y.presheaf.map_comp, ← Y.presheaf.map_comp]
+    congr 1
+#align algebraic_geometry.PresheafedSpace.is_open_immersion.pullback_cone_of_left_condition AlgebraicGeometry.PresheafedSpace.IsOpenImmersion.pullback_cone_of_left_condition
+
+/-- We construct the pullback along an open immersion via restricting along the pullback of the
+maps of underlying spaces (which is also an open embedding).
+-/
+def pullbackConeOfLeft : PullbackCone f g :=
+  PullbackCone.mk (pullbackConeOfLeftFst f g) (Y.ofRestrict _)
+    (pullback_cone_of_left_condition f g)
+#align algebraic_geometry.PresheafedSpace.is_open_immersion.pullback_cone_of_left AlgebraicGeometry.PresheafedSpace.IsOpenImmersion.pullbackConeOfLeft
+
+variable (s : PullbackCone f g)
+
+/-- (Implementation.) Any cone over `cospan f g` indeed factors through the constructed cone.
+-/
+def pullbackConeOfLeftLift : s.pt ⟶ (pullbackConeOfLeft f g).pt where
+  base :=
+    pullback.lift s.fst.base s.snd.base
+      (congr_arg (fun x => PresheafedSpace.Hom.base x) s.condition)
+  c :=
+    { app := fun U =>
+        s.snd.c.app _ ≫
+          s.pt.presheaf.map
+            (eqToHom
+              (by
+                dsimp only [Opens.map, IsOpenMap.functor, Functor.op]
+                congr 2
+                let s' : PullbackCone f.base g.base := PullbackCone.mk s.fst.base s.snd.base
+                  -- Porting note : in mathlib3, this is just an underscore
+                  (congr_arg Hom.base s.condition)
+
+                have : _ = s.snd.base := limit.lift_π s' WalkingCospan.right
+                conv_lhs =>
+                  erw [← this]
+                  dsimp
+                  -- Porting note : need a bit more hand holding here about function composition
+                  rw [coe_comp, show ∀ f g, f ∘ g = fun x => f (g x) from fun _ _ => rfl]
+                  erw [← Set.preimage_preimage]
+                erw [Set.preimage_image_eq _
+                    (TopCat.snd_openEmbedding_of_left_openEmbedding hf.base_open g.base).inj]
+                rfl))
+      naturality := fun U V i => by
+        erw [s.snd.c.naturality_assoc]
+        rw [Category.assoc]
+        erw [← s.pt.presheaf.map_comp, ← s.pt.presheaf.map_comp]
+        congr 1 }
+#align algebraic_geometry.PresheafedSpace.is_open_immersion.pullback_cone_of_left_lift AlgebraicGeometry.PresheafedSpace.IsOpenImmersion.pullbackConeOfLeftLift
+
+-- this lemma is not a `simp` lemma, because it is an implementation detail
+theorem pullbackConeOfLeftLift_fst :
+    pullbackConeOfLeftLift f g s ≫ (pullbackConeOfLeft f g).fst = s.fst := by
+  -- Porting note : `ext` did not pick up `NatTrans.ext`
```

Comment:
Could you add the missing `ext` lemma?

Reply:
I tired `attribute [ext] NatTrans.ext`, still does not work. After quick searching porting note, `ext` regression is not an isolated case. @semorrison

Reply:
The issue is that current `ext` doesn't see through type synonyms. I've proposed a fix but it's unclear if we'll actually implement it, so in the meantime the solution is to add an extra `ext` lemma, essentially for the morphisms of every new category that gets defined. Does that make sense?

Reply:
See https://github.com/leanprover-community/mathlib4/pull/5191